src/amici.cpp

```diff
@@ -1383,6 +1384,375 @@ int getDiagnosisB(int it, void *ami_mem, UserData *udata, ReturnData *rdata, Tem
+     * @param[out] rdata pointer to the return data struct @type ReturnData
+     * @param[in] newton integer flag indicating the run of the Newton solver
+     * @param[in] clock_t starttime timer running from the start of Newton solver
+     * @return int status flag indicating success of execution @type int
+     */
+    
+    realtype *x_tmp;
+    
+    // Get time for Newton solve
+    rdata->newton_time[0] = run_time;
+    
+    // Since the steady state was found, current time is set to infinity
+    tdata->t = INFINITY;
+    
+    // Write output
+    x_tmp = N_VGetArrayPointer(tdata->x);
+    for (int ix=0; ix<udata->nx; ix++) {
+        rdata->xss[ix] = x_tmp[ix];
+    }
+    
+    // Write flag for the Newton solver
+    *rdata->newton = (double) newton;
+    
+    return(AMICI_SUCCESS);
+}
+
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+
+int getNewtonSimulation(void *ami_mem, UserData *udata, TempData *tdata, ReturnData *rdata) {
+    /**
+     * getNewtonSimulation solves the forward problem, if the first Newton solver run did not succeed.
+     *
+     * @param[in] udata pointer to the user data struct @type UserData
+     * @param[in] tdata pointer to the temporary data struct @type TempData
+     * @param[out] rdata pointer to the return data struct @type ReturnData
+     * @param[in] newton integer flag indicating the run of the Newton solver
+     * @param[in] clock_t starttime timer running from the start of Newton solver
+     * @return int status flag indicating success of execution @type int
+     */
+    
+    
+    double res_abs;
+    double res_rel;
+    double sim_time;
+    int status = (int) *rdata->status;
+    realtype *x_tmp;
+    N_Vector rel_x = N_VNew_Serial(udata->nx);
+    N_Vector tmp_x = N_VNew_Serial(udata->nx);
+    
+    /* Newton solver did not work, so try a simulation */
+    if (tdata->t >= 1e6) {
+        sim_time = 10.0*(tdata->t);
+    } else {
+        sim_time = 1e6;
+    }
+    status = AMISolve(ami_mem, RCONST(sim_time), tdata->x, tdata->dx, &(tdata->t), AMICI_NORMAL);
+    
+    if (status == AMICI_SUCCESS) {
+        /* Check residuals */
+        res_abs = sqrt(N_VDotProd(tdata->xdot, tdata->xdot));
+    
+        /* Ensure positivity for relative residual */
+        N_VScale(1.0, tdata->x, tmp_x);
+        N_VAbs(tmp_x, tmp_x);
+        x_tmp = N_VGetArrayPointer(tmp_x);
+        for (int ix=0; ix<udata->nx; ix++) {
+            if (x_tmp[ix] < udata->atol) {
+                x_tmp[ix] = udata->atol;
+            }
+        }
+        N_VDiv(tdata->xdot, tmp_x, rel_x);
+        res_rel = sqrt(N_VDotProd(rel_x, rel_x));
+    
+        /* residuals are small? */
+        if (res_abs < udata->atol || res_rel < udata->rtol) {
```

Comment:
That's what I've actually planned: I wanted to define an event, which is triggered if xdot is small. The problem: The event needs to be added to the other events before compiling...
That is maybe not super-nice and functionality of the Newton solver would be depending on the functionality of the events forever, so I decided against it.
A loop with single-steps might be a neat idea, I also thought about it and is probably the way to go. But this, I wanted to discuss with you offline in more detail before implemented. Right now, the merge ist more important and the current implementation works fine.